grassroots-backend/src/phone-canvass/Scheduler/PhoneCanvassScheduler.ts

```diff
@@ -0,0 +1,153 @@
+/* eslint-disable grassroots/entity-use */
+
+import { filter, firstValueFrom, Observable, Subject } from "rxjs";
+import { PhoneCanvassContactEntity } from "../entities/PhoneCanvassContact.entity.js";
+import {
+  InProgressCall as InProgressCall,
+  CompletedCall,
+  NotStartedCall,
+  RingingCall as RingingCall,
+  InitiatedCall,
+} from "./PhoneCanvassCall.js";
+import { PhoneCanvassMetricsTracker as PhoneCanvassMetricsTracker } from "./PhoneCanvassMetricsTracker.js";
+import { PhoneCanvassSchedulerStrategy } from "./Strategies/PhoneCanvassSchedulerStrategy.js";
+import { NoOvercallingStrategy } from "./Strategies/NoOvercallingStrategy.js";
+import { CallStatus } from "grassroots-shared/dtos/PhoneCanvass/CallStatus.dto";
+
+export interface Caller {
+  id: number;
+  availabilityStartTime: number /* Relative to Date.now()*/;
+}
+
+export interface PhoneCanvassScheduler {
+  readonly calls: Observable<NotStartedCall>;
+  start(): Promise<void>;
+  stop(): void;
+  addCaller(id: number): void;
+  removeCaller(id: number): void;
+  waitForIdleForTest(): Promise<void>;
+  getNextIdleCallerId(): number | undefined;
+}
+
+export class PhoneCanvassSchedulerImpl implements PhoneCanvassScheduler {
+  #callsObservable = new Subject<NotStartedCall>();
+  // Clients subscribe to this list of calls which need to be placed.
+  readonly calls = this.#callsObservable.asObservable();
+
+  #strategy: PhoneCanvassSchedulerStrategy;
+
+  callsByStatus = {
+    NOT_STARTED: new Map<number, NotStartedCall>(),
+    QUEUED: new Map<number, RingingCall>(),
+    INITIATED: new Map<number, InitiatedCall>(),
+    RINGING: new Map<number, RingingCall>(),
+    IN_PROGRESS: new Map<number, InProgressCall>(),
+    COMPLETED: new Map<number, CompletedCall>(),
+  } as const satisfies Record<CallStatus, unknown>;
+  // From caller id.
+  #callers = new Map<number, Caller>();
+
+  #running = false;
+  #pendingContacts: PhoneCanvassContactEntity[];
+
+  metricsTracker: PhoneCanvassMetricsTracker;
+
+  getCurrentTime(): number {
+    return Date.now();
+  }
+
+  constructor(contacts: PhoneCanvassContactEntity[]) {
+    this.#pendingContacts = contacts.filter((contact) => {
+      // TODO(MVP): should this include "UNABLE_TO_CONTACT"?
+      return contact.callStatus === "NOT_STARTED";
+    });
+    this.metricsTracker = new PhoneCanvassMetricsTracker();
+    this.#strategy = new NoOvercallingStrategy(this.metricsTracker);
+  }
+
+  async start(): Promise<void> {
+    if (this.#running) {
+      throw new Error("Call to startScheduler when scheduler already running.");
+    }
+    this.#running = true;
+    while (this.#running) {
+      await this.#strategy.waitForNextCall();
```

Comment:
because of the await, I believe `stop` will allow one last call to go through. Maybe not a big deal, but we could check after the await `if (!this.#running) break;`

Reply:
Great point. eslint doesn't realize that the value could change during that await, so I needed to
```// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition```
which is a bit scary.